ipnsfs/dir.go

```diff
@@ -0,0 +1,343 @@
+		if lnk.Name == name {
+			nd, err := lnk.GetNode(d.fs.dserv)
+			if err != nil {
+				return nil, err
+			}
+			i, err := ft.FromBytes(nd.Data)
+			if err != nil {
+				return nil, err
+			}
+
+			switch i.GetType() {
+			case ufspb.Data_Directory:
+				ndir := NewDirectory(name, nd, d, d.fs)
+				d.childDirs[name] = ndir
+				return ndir, nil
+			case ufspb.Data_File:
+				return nil, fmt.Errorf("%s is not a directory", name)
+			case ufspb.Data_Metadata:
+				return nil, ErrNotYetImplemented
+			default:
+				return nil, ErrInvalidChild
+			}
+		}
+
+	}
+
+	return nil, ErrNoSuch
+}
+
+func (d *Directory) Child(name string) (FSNode, error) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	dir, err := d.childDir(name)
+	if err == nil {
+		return dir, nil
+	}
+	fi, err := d.childFile(name)
+	if err == nil {
+		return fi, nil
+	}
+
+	return nil, ErrNoSuch
+}
+
+func (d *Directory) List() []string {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	var out []string
+	for _, lnk := range d.node.Links {
+		out = append(out, lnk.Name)
+	}
+	return out
+}
+
+func (d *Directory) Mkdir(name string) (*Directory, error) {
+	d.lock.Lock()
+
+	_, err := d.childDir(name)
+	if err == nil {
+		d.lock.Unlock()
+		return nil, errors.New("directory by that name already exists")
+	}
+	_, err = d.childFile(name)
+	if err == nil {
+		d.lock.Unlock()
+		return nil, errors.New("file by that name already exists")
+	}
+
+	ndir := &dag.Node{Data: ft.FolderPBData()}
+	err = d.node.AddNodeLinkClean(name, ndir)
+	if err != nil {
+		d.lock.Unlock()
+		return nil, err
+	}
+	d.lock.Unlock()
+
+	err = d.parent.closeChild(d.name, d.node)
+	if err != nil {
+		return nil, err
+	}
+
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	return d.childDir(name)
+}
+
+func (d *Directory) Unlink(name string) error {
+	d.lock.Lock()
+	delete(d.childDirs, name)
+	delete(d.files, name)
+
+	err := d.node.RemoveNodeLink(name)
+	if err != nil {
+		d.lock.Unlock()
+		return err
+	}
+	d.lock.Unlock()
+
+	return d.parent.closeChild(d.name, d.node)
+}
+
+// RenameEntry renames the child by 'oldname' of this directory to 'newname'
+func (d *Directory) RenameEntry(oldname, newname string) error {
+	// Is the child a directory?
+	dir, err := d.childDir(oldname)
+	if err == nil {
+		dir.name = newname
+
+		err := d.node.RemoveNodeLink(oldname)
+		if err != nil {
+			return err
+		}
+		err = d.node.AddNodeLinkClean(newname, dir.node)
+		if err != nil {
+			return err
+		}
+
+		delete(d.childDirs, oldname)
+		d.childDirs[newname] = dir
+		return d.parent.closeChild(d.name, d.node)
+	}
+
+	// Is the child a file?
+	fi, err := d.childFile(oldname)
+	if err == nil {
+		fi.name = newname
+
+		err := d.node.RemoveNodeLink(oldname)
+		if err != nil {
+			return err
+		}
+
+		nd, err := fi.GetNode()
+		if err != nil {
+			return err
+		}
+
+		err = d.node.AddNodeLinkClean(newname, nd)
+		if err != nil {
+			return err
+		}
+
+		delete(d.childDirs, oldname)
+		d.files[newname] = fi
+		return d.parent.closeChild(d.name, d.node)
```

Comment:
there's a lot of common code in this section and the one above. if someone comes in and changes the semantics of one, they may fail to change the other. the whole inner part of these ifs is the same except one line